presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -59,98 +52,102 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
 
     public Optional<DecorrelatedNode> decorrelateFilters(PlanNode node, List<Symbol> correlation)
     {
-        PlanNodeSearcher filterNodeSearcher = searchFrom(node, lookup)
-                .where(FilterNode.class::isInstance)
-                .recurseOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class));
-        List<FilterNode> filterNodes = filterNodeSearcher.findAll();
+        // TODO: when correlations list empty this should return immediately. However this isn't correct
+        // right now, because for nested subqueries correlation list is empty while there might exists usages
+        // of the outer most correlated symbols
+
+        Optional<DecorrelationResult> decorrelationResultOptional = lookup.resolve(node).accept(new DecorrelatingVisitor(correlation), null);
+        return decorrelationResultOptional.flatMap(decorrelationResult -> decorrelatedNode(
+                decorrelationResult.correlatedPredicates,
+                decorrelationResult.node,
+                correlation));
+    }
 
-        if (filterNodes.isEmpty()) {
-            return decorrelatedNode(ImmutableList.of(), node, correlation);
-        }
+    private class DecorrelatingVisitor
+            extends PlanVisitor<Optional<DecorrelationResult>, Void>
+    {
+        final List<Symbol> correlation;
 
-        if (filterNodes.size() > 1) {
-            return Optional.empty();
+        DecorrelatingVisitor(List<Symbol> correlation)
+        {
+            this.correlation = correlation;
         }
 
-        FilterNode filterNode = filterNodes.get(0);
-        Expression predicate = filterNode.getPredicate();
-
-        if (!isSupportedPredicate(predicate)) {
-            return Optional.empty();
+        @Override
+        protected Optional<DecorrelationResult> visitPlan(PlanNode node, Void context)
+        {
+            return Optional.of(new DecorrelationResult(
```

Comment:
There is check in `PlanNodeDecorrelator#decorrelatedNode` that makes sure that all correlation are removed. I think it's safer to have such check `decorrelatedNode`.

This means that `DecorrelationResult` is rather "the best we can do" but please verify later if we managed to decorrelate everything.

Reply:
When you are pushing symbols up and you see `LIMIT n` you might need to cut the process there because it might not be possible to remove `LimitNode`. If you left `LimitNode` then you would end up with invalid query. Remember, some nodes below are already decorrelated so technically there might be no more correlated symbols in subquery. 
Similar situation with `AggregtaionNode`.